### PR TITLE
Add withLogConsumer + general logging improvements

### PR DESCRIPTION
**Pull request by cristianrgreco**

### CODE REVIEW

File: src/logger.ts [hunk 1]
```diff
@@ -2,79 +2,56 @@
...
-  public enabled(): boolean {
-    return true;
+  public trace(message: Message, options?: Options): void {
+    this.logger(`${this.showLevel ? "[TRAC] " : ""}${this.renderOptions(options)}${message}`);
   }
 
-  public trace(message: Message): void {
-    this.traceLogs.push(message);
+  public debug(message: Message, options?: Options): void {
+    this.logger(`${this.showLevel ? "[DEBU] " : ""}${this.renderOptions(options)}${message}`);
```
[the remark this answers is not in the document]
Comment (AlexPtnn) [reply src/logger.ts:26]: DEBUG isntead of DEBU

File: src/logger.ts [hunk 2]
```diff
@@ -2,79 +2,57 @@ import debug, { IDebugger } from "debug";
...
 
-  public error(message: Message): void {
-    this.errorLogs.push(message);
+  private renderOptions(options?: Options): string {
+    let str = "";
+    if (options?.containerId) {
+      str += `[${options.containerId.substring(0, 12)}] `;
+    }
+    if (options?.imageName) {
+      str += `[${options.imageName}] `;
```
Comment (AlexPtnn) [inline src/logger.ts:47]: Can you set here name of container? And if it is not exists use imageName? Name will be shorter than imageName in output.

File: src/generic-container/generic-container.ts [hunk 3]
```diff
@@ -203,10 +206,18 @@ export class GenericContainer implements TestContainer {
...
+    if (containerLog.enabled() || this.logConsumer !== undefined) {
+      const logStream = await containerLogs(container);
+
+      if (this.logConsumer !== undefined) {
+        this.logConsumer(logStream);
+      }
+
+      if (containerLog.enabled()) {
+        logStream
+          .on("data", (data) => containerLog.trace(data.trim(), { containerId: container.id }))
```
Comment (AlexPtnn) [inline src/generic-container/generic-container.ts:218]: Will be nice to have containerName/imageName here (one of, name > image), as we discussed in my PR

File: src/logger.ts (hunk 2 above)
Comment (cristianrgreco) [reply src/logger.ts:47]: This field is only set when pulling an image, at which point there isn't a container

File: src/generic-container/generic-container.ts (hunk 3 above)
Comment (cristianrgreco) [reply src/generic-container/generic-container.ts:218]: Actually see this comment on your PR where I explained why not to use the container name: https://github.com/testcontainers/testcontainers-node/pull/558#issuecomment-1536068467

**Pull request merged**